### ISSUE TITLE
DefaultS3StoreImpl doesn't allow dynamic buckets

### ISSUE THREAD

**Issue by nuvitong** [paulcwarren/spring-content#22]
Hi Paul,

first of all thanks for this very nice piece of work.

In our use case we have not only one bucket for our s3 storage, but several, namely one for each customer. So we would like to set the customer id as property of the `@Entity` and use this *customer id* as bucket name. As far as I can see, this is not possible with the `DefaultS3StoreImpl`, as thie private method `absolutify()` uses the `bucket` of the `DefaultS3StoreImpl` within the storage path.

What do you think would be the best way, to use a property of the entity as bucket name? Should `absolutify()` be `protected`, so it could be overwritten accordingly? 

What is your opinion on that?

Kind regards,

Tobias

**Comment by paulcwarren** [comment]
Hi Tobias,
You are correct it is not possible with the current DefaultS3StoreImpl.  The easiest thing to do for the code is to make absolutify protected , as you say.  That would mean that you would also have to create your own subclass of S3StoreFactoryBean that returns your subclass implementation and plug that factory into your application specifying the `storeFactoryBeanClass` on your `EnableS3Stores` annotation.  

The other thing we could do is add a Configurer API to the s3 module allowing the contribution a BuckerResolver bean.  Something like the following:-

public interface BucketResolver<S> {
  String resolveBucket(S entity); 
}

In one of your @Configuration classes this would allow you to contribute a bean implementing this interface that return your entity's customerId.  

With both options I guess they would to ensure the bucket exists creating if needed.   They would need the `amazonClient` bean therefore.  If that customer ever became an x customer I dont think there is anything we could do to remove these buckets again so we would probably just have to leave orphaned buckets and perhaps have purging performed by your application as a separate service.  

Some early thoughts.  Anyway, the ask is perfectly reasonable.  We have suspected folks would want to do this.  So we are happy to put one of these options in, or accept a PR of course.  Do you have any preference?

_Paul

**Comment by nuvitong** [comment]
Hi Paul,

thanks for the quick response.

After a second thought and your explanation, making `absolutify` *protected* and overwrite it looks more as a workaround to me. Too much unneccessary work for just using dynamic buckets.

I thought about your second option of a BucketResolver bean. I guess this would work for using in combination with the `ContentStore` interface, where the methods know the entity. Nonetheless the `Store` interface only knows the *SID* (`getResource(SID)`). So, as far as I can see the BucketResolver could not be used with the `Store`.

Currently I have a third option in mind: if it would be allowed, to leave the `spring.content.s3.bucket` empty, so that it would not be taken into account, then the developer could specify an accordant SID, which would contain the bucket as well. 

```java
public class ContentId implements Serializable {
    private String bucket;
    private String fileId;
    // getters and setters ...
}
```
Perhaps it would be neccessary to implement an accordant conversion service to create the `location` of the file, but that would be totally fine from my point of view.

About your point with missing and orphaned buckets: i already thought about that and would implement an accordant `@StoreEventHandler` with an `@HandleBeforeSetContent` (creating the bucket, if none exists) and an `@HandleAfterUnsetContent` (deleting the bucket, if that was the last content within the bucket).

What do you think?

Kind regards,

Tobias

**Comment by paulcwarren** [comment]
Hi Tobias,

Thanks for raising `Store`.  When I replied I hadn't considered Store but we absolutely should.  

I like the idea of using the first element of the path as the bucket.  But I don't particularly like the idea of using a path-based resolution for Store but using a BucketResolver for ContentStore.  It would be nice to have the same behavior across all types of Store.   We could use path-based resolution for ContentStore as well (ContentStores will only assign `@ContentId` with a UUID if the field is `null`) but then your application code would need to pre-populate the `@ContentId` field with a value like `<customer_id>/<uuid>` when new entities are created.   This would have to be on a pre-save event handler or in an overridden repository save implementation.   At first I thought this might be undue burden on the developer but when I thought about it a bit more I am not sure there is actually much to choose between putting `@ContentId` setting login in/before `save` versus bucket resolution login in a `BucketResolver`.

What say you?

To summarize this behavior across the store types I think it would look something like this:

```

			@ContentId	BUCKET		Method			Bucket			Lookup
Store			—		set		getResource /a/b	<BUCKET>		<BUCKET>/a/b		
Store 			—		unset		getResource /a/b	a 			/a/b
ContentStore		null		set		setContent 		<BUCKET>		<BUCKET>/uuid		
ContentStore		null		unset		setContent 		error
ContentStore		<user_id>	set		setContent 		<BUCKET>		<BUCKET>/user_id	
ContentStore		<user_id>	unset		setContent 		error		
ContentStore		<user path>	set		setContent		<BUCKET>		<BUCKET>/a/b 		
ContentStore		<user path>	unset		setContent		a			/a/b 				
``` 

NB: getResource would presumably be followed by a getOutputStream and a wrote operation.

**Comment by paulcwarren** [comment]
Hi Tobias,

Thanks for bringing up `Store`.  When I replied initially I had not considered that but we absolutely should.  This behavior needs to be handled uniformly across all types of store.

I like where you are going with alternative 3.  I like that it leverages the Store's generic ID and the fact that it re-enforces the bucket as part of the content's ID.  I am not sure if you meant for `ContentId` to be part of the Spring Content API, or not.   I know several of the Spring Data's went this route.  Mongo has its `ObjectId` and Cassandra has its `MapId`.  Under this alternative the application would need to "pre-populate" the Entity's content id field using an overridden repository save (or equivalent).  
 
An alternative, on the same theme (alternative 3b I guess), would be to leverage the conversion service to extract the bucket and the id/path.  Initially I liked this as it would work for any user-defined ID type and SC S3 wouldn't need to define one.  :
```
        import com.amazonaws.services.s3.model.Bucket;
        ...
	@Override
	public Resource getResource(SID id) {
		Bucket bucket = converter.convert(id, Bucket.class);
		String objectId = converter.convert(id, String.class);
		location = absolutify(bucket,fileId);
		Resource resource = loader.getResource(location);
		return resource;
	}
``` 
But this is also defining a content id type "in effect" and it is a little bit of an abuse of a conversion service.  So I think your original alternative #3 is probably the simplest and cleanest.

**Comment by paulcwarren** [comment]
Fixed in commit e0a3e377379aa0e28c8f04917388398c467d5a95.  

As discussed - introduced an S3ContentId object that can be used to capture the bucket name.  Let me know if you have any issues.  Will document in reference guide in next commit.

**Comment by paulcwarren** [comment]
That's some of the downsides of the content id object approach.  

Another alternative would be to introduce an Spring Content S3-specific annotation `@Bucket`.  This could be used to annotate the entity field that should be used as the bucket.  This would allow you to mark your customer id field as the bucket field.  This would only work for `AssociativeStore` and `ContentStore` - but maybe that's ok?  It would also be specific to 'S3' too.  But this approach is definitely in keeping with Spring Data extension model.  If we were to make this the behavior for those Store's then I would assume that `Store` would determine the bucket from either the AWS_BUCKET env var or the first element in the resource's location (or error if that cant be determined).   Or we could just leave single bucket behavior and wait for a customer to actually come tell us we need it under the general auspice of YAGNI.

**Comment by nuvitong** [comment]
Funny, I also came along with the `@Bucket` annotation for my Entity. A second later I remembered, that it is the `SID`, which has to know the bucket name, as I would like to use the `Store`-Interface. 

When using the current 0.0.11-SNAPSHOT I figured out, that the `S3ContentId` has another drawback, if the S3StoreImpl should create a new *ObjectId* (within `setContent()`).

Currently I'm trying to use a `ContentIdResolver`-Bean, which would have to be implemented, if the `SID` contains the bucket. I'll try to provide a PR on the weekend, so that we could discuss this option on the code.

**Comment by paulcwarren** [comment]
You want to use `Store`.  Ok.  Because you had an Entity model I had it in my head you were using `ContentStore` for some reason.

Then I think we should also investigate whether Store should be `Store.getResource(S)` rather than `Store.getResource(SID)`.  That way SC provides an API where a Resource can be loaded by passing the Entity, rather than an its ID.  This is actually pretty neat at the java API level as it allows us to expose the full set of Spring Content functionality through a single `Store` interface.  So this would allow you to, for example, use `myStore.getResource(myEntity)` to fetch an entitiy's resource.  The loaded resource would have a reference to the entity and therefore internally in the `DefaultS3StoreImpl` we would be able make the entity->resource associations, manage the entity's content annotations - just like we do in the `ContentStore` methods, etc...I think...

At the REST layer this would mean that logically an Entity's content is then addressable via the Entity's URL rather than via its own content id.  We would have a set of SC controllers whose mappings overlapped exactly with those of Spring Data's - which definitely causes issues.  That said we kinda have that situation today with `ContentEntityRestController` anyway and we recommend deconflicting those through a different `StoreRestResource` path.  So Spring Data entities are address at `/documents` but their content is addressed at `/documentsContent`.  That would most likely work here too.

I think `Renderable` extensions would be fine.  But I will need to have a think/look at Searchable (our fulltext indexing).

Anyways, a PR would be good as it adds meat to the bones.  I will try and spike on this suggestion too on a wip branch.

**Comment by nuvitong** [comment]
Hi Paul,

I finally found the time to provide the promised PR. I named the addressed Interface `S3ContentIdHelper` although I don't like the name very much.

The interface provides methods to get the bucket name and the object id of the content id. Additionally there is a method to validate the content id. 

If the developer would like to use a custom type as content id, she has to implement this interface in order to get the bucket name and object id from the content id object. If there should be a special validation for the content id, this could be implemented in `validate(SID)`.

Currently I did not add the configuration for adding a custom `S3ContentIdHelper` within the `S3StoreFactoryBean`, but I thing you get the point of my idea.

What do you think?

**Comment by paulcwarren** [comment]
Thanks Tobias, I am at a conference this week.  I will try and look at this if I get time but I just wanted to provide forewarning in case I don't.

**Comment by paulcwarren** [comment]
Hey Tobias, thanks for the PR.  

Just in case you hadn't spotted this.  "range" requests *are* supported by the `ContentStore`.  See [this](https://github.com/paulcwarren/spring-content/blob/master/spring-content-rest/src/main/java/internal/org/springframework/content/rest/controllers/ContentEntityRestController.java#L67) and [this](https://github.com/paulcwarren/spring-content/blob/014944018c1b47706fc8cdf92ead420bfe47474c/spring-content-rest/src/main/java/internal/org/springframework/content/rest/controllers/ContentPropertyRestController.java#L70).   These are the controllers that will get called when your interfaces extend `ContentStore` and you make an http byte-range request on those interfaces.   

I don't know if that makes any difference to your proposal?  

Assuming not.  I do like the proposal though.  I agree `S3ContentIdHelper` is not a great name (is `S3ContentIdResolver` any better?).   It's a nice and clean "get out of jail free" option.  But I also think from a developer's perspective the `@Bucket` annotation for use with the `AssociativeStore` and `ContentStore` is the most usable .  So I am wondering if we can do both here with a default resolver.  You would be free to set your own resolver but the default behavior would be to use @Bucket annotation/env var.

Perhaps, I'll try and update the code so we can see how that would look.

**Comment by paulcwarren** [comment]
Hey Tobias, I did a bit of work to tidy this up today (including updating the docs).  Can you check whether this still satisfies you requirements?  If so I will merge this back into master and cut a release next week.

**Comment by nuvitong** [comment]
Hi Paul,

thanks for your great work on this issue. At a first glance your changes look very promising to me. I will check on monday at work, if I find any pitfalls regarding our needs. 

Kind regards,

Tobias

**Comment by paulcwarren** [comment]
Hi Tobias,
I am assuming this is now resolved for you and closing.  Please just shout if not, or if there is anything else.

**Comment by nuvitong** [comment]
Hi Paul,

sorry for the late reply. It was very busy at work, before I was allowed to go on vacation. :-)

Indeed the issue is resolved for me. I'm very happy to use spring-content and will try to contribute, if I need additional features.

Kind regards,

Tobias

**Comment by paulcwarren** [comment]
Thanks Tobias.  Glad we resolved it satisfactorily.  Don't hesitate moving forwards.  Great to work with you.
